Allow different version encoding for custom builds

mnagy opened the pull request:
This PR adds a new `buildApiVersion` parameter to the custom build strategy. The version that is specified is then used to serialize the Build object that is passed into the custom builder via the BUILD environment variable. Omitting the parameter will use the same version as it is using now (the latest one).

Trello card: https://trello.com/c/2CrsqGjY/638-5-do-not-pass-internal-build-object-json-serialization-to-builder-container

mnagy commented:
This PR needs approval from @openshift/api-review

smarterclayton commented on pkg/build/api/v1beta3/types.go:
```diff
@@ -308,6 +308,9 @@ type CustomBuildStrategy struct {
 	Secrets []SecretSpec `json:"secrets,omitempty" description:"a list of secrets to include in the build pod in addition to pull, push and source secrets"`
+
+	// BuildAPIVersion is the requested API version for the Build object serialized and passed to the custom builder
+	BuildAPIVersion string `json:"buildApiVersion,omitempty" description:"requested API version for the Build object serialized and passed to the custom builder"`
```
The JSON name here should be `buildAPIVersion`

mfojtik commented on pkg/build/controller/strategy/custom_test.go:
```diff
@@ -104,6 +106,33 @@ func TestCustomCreateBuildPodExpectedForcePull(t *testing.T) {
+		Codec: latest.Codec,
+	}
+
+versionLoop:
```
nasty

mnagy commented:
Changed JSON name to `buildAPIVersion` and addressed Michal's comment.

smarterclayton commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -22,12 +23,22 @@ type CustomBuildStrategy struct {
+	strategy := build.Spec.Strategy.CustomStrategy
+
+	codec := bs.Codec
+	if strategy.BuildAPIVersion != "" {
```
This should be `len(strategy.BuildAPIVersion) != 0` which is our convention.

smarterclayton commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -22,12 +23,22 @@ type CustomBuildStrategy struct {
+	if strategy.BuildAPIVersion != "" {
+		interfaces, err := latest.InterfacesFor(strategy.BuildAPIVersion)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get codec for requested API version (%q): %v", strategy.BuildAPIVersion, err)
```
This message needs to indicate that the custom build has a buildAPIVersion specified which does not exist - the error is too generic.  Where is this error reported to the user?

mnagy replied:
The created Build will get stuck in New phase:
```yaml
status:
  config:
    kind: BuildConfig
    name: ruby-sample-build
    namespace: demo
  message: 'Failed to create a build pod spec for build demo/ruby-sample-build-9:
    failed to get codec for requested API version ("v1beta2"): unsupported storage
    version: v1beta2 (valid: v1, v1beta3).'
  outputDockerImageReference: 172.30.57.7:5000/demo/origin-ruby-sample:latest
  phase: New
  reason: CannotCreateBuildPodSpec
```
Is that good enough?

deads2k replied:
> Is that good enough?

Can I mutate the build to fix the problem?  If not, why wouldn't it go into `Failed`?

mnagy replied:
No, you can't. It won't go into `Failed` because the retry controller doesn't recognize it as a fatal error. We can change that. How about adding BC validation check for the version, making sure we recognize it?

deads2k replied:
> No, you can't. It won't go into Failed because the retry controller doesn't recognize it as a fatal error. We can change that. How about adding BC validation check for the version, making sure we recognize it?

Ok, this seems pretty fatal.  You can't update it and the build will never progress.

Adding it to BC validation might be nice, but it won't be sufficient.  Consider cases where you've requested an API version that used to exist, but has since been removed.  Validation passed last week, but you can still end up in this case.

mnagy replied:
OK, will update to recognize the error as fatal and add BC validation for better user experience.

smarterclayton replied:
BC validation being risky is a problem solved by having the storage upgrade
path which needs to get turned into a card so we can start sticking things
in it (correcting data in objects to the newest version).  If we drop v1,
we're going to have to touch every object in the system to set it to v2, so
we can update the BC at that time.

On Fri, Jan 22, 2016 at 9:31 AM, Martin Nagy <notifications@github.com>
wrote:

> In pkg/build/controller/strategy/custom.go
> <https://github.com/openshift/origin/pull/6760#discussion_r50542087>:
>
> > @@ -22,12 +23,22 @@ type CustomBuildStrategy struct {
> >
> >  // CreateBuildPod creates the pod to be used for the Custom build
> >  func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod, error) {
> > -	data, err := bs.Codec.Encode(build)
> > +	strategy := build.Spec.Strategy.CustomStrategy
> > +
> > +	codec := bs.Codec
> > +	if strategy.BuildAPIVersion != "" {
> > +		interfaces, err := latest.InterfacesFor(strategy.BuildAPIVersion)
> > +		if err != nil {
> > +			return nil, fmt.Errorf("failed to get codec for requested API version (%q): %v", strategy.BuildAPIVersion, err)
>
> OK, will update to recognize the error as fatal and add BC validation for
> better user experience.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6760/files#r50542087>.
>

mnagy replied:
Trello'd here: https://trello.com/c/xxdyGxp1/850-provide-a-storage-upgrade-path-for-objects

mnagy replied:
I wonder, should I also make any change to UPGRADING.md with respect to this change?

liggitt commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -22,12 +23,22 @@ type CustomBuildStrategy struct {
 
 // CreateBuildPod creates the pod to be used for the Custom build
 func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod, error) {
-	data, err := bs.Codec.Encode(build)
```
was this previously encoding the internal version? that should be fixed regardless of whether strategy.BuildAPIVersion is set.

mnagy replied:
Yes, all strategies have their codecs set in run_components.go:
https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/run_components.go#L190-L191
https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/run_components.go#L203-L218
Should we just use latest?

smarterclayton replied:
It needs to be coming from the storage version config flag.

On Fri, Jan 22, 2016 at 1:51 PM, Martin Nagy <notifications@github.com>
wrote:

> In pkg/build/controller/strategy/custom.go
> <https://github.com/openshift/origin/pull/6760#discussion_r50574745>:
>
> > @@ -22,12 +23,22 @@ type CustomBuildStrategy struct {
> >
> >  // CreateBuildPod creates the pod to be used for the Custom build
> >  func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod, error) {
> > -	data, err := bs.Codec.Encode(build)
>
> Yes, all strategies have their codecs set in run_components.go:
>
> https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/run_components.go#L190-L191
>
> https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/run_components.go#L203-L218
> Should we just use latest?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6760/files#r50574745>.
>

mnagy replied:
It is.

mnagy commented:
Addressed comments. The build will now fail instantly. Example build status:
```yaml
status:
  completionTimestamp: 2016-01-26T11:02:43Z
  config:
    kind: BuildConfig
    name: ruby-sample-build
    namespace: demo
  message: 'Failed to create a build pod spec for build demo/ruby-sample-build-17:
    invalid buildAPIVersion specified in custom build strategy ("v1beta2"): unsupported
    storage version: v1beta2 (valid: v1, v1beta3).'
  outputDockerImageReference: 172.30.57.7:5000/demo/origin-ruby-sample:latest
  phase: Failed
  reason: CannotCreateBuildPodSpec
```

mnagy commented:
PTAL

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -33,13 +33,16 @@ const maxRetries = 60
-		build.Status.Reason = buildapi.StatusReasonExceededRetryTimeout
+		if !isFatal {
+			build.Status.Reason = buildapi.StatusReasonExceededRetryTimeout
+		}
```
so we don't set any reason if it's a fatal error?  or you're assuming that was done elsewhere?

bparees commented:
lgtm, @openshift/api-review bump.

smarterclayton commented:
API approved

bparees commented:
[merge]

bparees commented:
_output/local/go/src/github.com/openshift/origin/pkg/build/controller/strategy/custom.go:30: cannot use strategy.BuildAPIVersion (type string) as type "k8s.io/kubernetes/pkg/api/unversioned".GroupVersion in argument to latest.InterfacesFor

smarterclayton commented:
Rebase gotcha

On Fri, Jan 29, 2016 at 6:07 PM, Ben Parees <notifications@github.com>
wrote:

> _output/local/go/src/
> github.com/openshift/origin/pkg/build/controller/strategy/custom.go:30:
> cannot use strategy.BuildAPIVersion (type string) as type "
> k8s.io/kubernetes/pkg/api/unversioned".GroupVersion in argument to
> latest.InterfacesFor
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6760#issuecomment-177011681>.
>

mnagy commented:
Fixed. @bparees can you re[merge]?

bparees commented:
@mnagy needs a squash and i'm not convinced those commits are what you intended since they are 11 days old.

bparees commented:
[test]

bparees commented:
[merge]
